Fix different seq length

Microve opened the pull request:
Differential Revision: D86685546




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

Microve commented:
topic: not user facing

eellison reviewed (approved):
Looks good, need to fix lint

Microve committed:
Fix different seq length in runtime_estimations_align_across_all_distributed_ranks (#167481)

Summary:

`align_runtime_estimations_across_all_distributed_ranks` is only needed when there are collectives in the graph. If there are no collectives, since Partitioner may make different decisions on the saved tensors. It could potentially cause failure in `runtime_estimations_align_across_all_distributed_ranks`

Test Plan:
Without this change, the job could fail because the different number of nodes in the backward graph:
https://www.internalfb.com/mlhub/pipelines/runs/mast/aps-aps-fb_him_sfsdp_h100-9990706fc5?job_attempt=0&version=0&tab=summary&env=PRODUCTION

tlparse of 16/3 in rank 113: https://fburl.com/3cxltwt4
tlparse of 16/3 in rank 124: https://fburl.com/p1grio09

With this change, it can run without hanging:
https://www.internalfb.com/mlhub/pipelines/runs/mast/aps-aps-fb_him_sfsdp_h100-c2323f01c2?job_attempt=0&version=0&tab=summary&env=PRODUCTION

Reviewed By: IvanKobzarev, eellison

Differential Revision: D86685546
